ms/MSOper/MSMetaData.cc

```diff
@@ -1132,61 +1132,53 @@ MDirection MSMetaData::getReferenceDirection(
 void MSMetaData::_getFieldsAndSpwMaps(
     std::map<Int, std::set<uInt> >& fieldToSpwMap,
     vector<std::set<Int> >& spwToFieldMap
-) {
+) const {
     // This method has the responsibility of setting _fieldToSpwMap and _spwToFieldIDMap
     if (! _fieldToSpwMap.empty() && ! _spwToFieldIDsMap.empty()) {
         fieldToSpwMap = _fieldToSpwMap;
         spwToFieldMap = _spwToFieldIDsMap;
         return;
     }
-    SHARED_PTR<Vector<Int> > allDDIDs = _getDataDescIDs();
-    SHARED_PTR<Vector<Int> > allFieldIDs = _getFieldIDs();
-    Vector<Int>::const_iterator endDDID = allDDIDs->end();
-    Vector<Int>::const_iterator curField = allFieldIDs->begin();
     fieldToSpwMap.clear();
     spwToFieldMap.resize(nSpw(True));
-    vector<uInt> ddidToSpwMap = getDataDescIDToSpwMap();
-    for (
-        Vector<Int>::const_iterator curDDID=allDDIDs->begin();
-        curDDID!=endDDID; ++curDDID, ++curField
-    ) {
-        uInt spw = ddidToSpwMap[*curDDID];
-        fieldToSpwMap[*curField].insert(spw);
-        spwToFieldMap[spw].insert(*curField);
-    }
-    std::map<Int, std::set<uInt> >::const_iterator mapEnd = fieldToSpwMap.end();
-    uInt mySize = 0;
-    for (
-        std::map<Int, std::set<uInt> >::const_iterator curMap = fieldToSpwMap.begin();
-        curMap != mapEnd; ++curMap
-    ) {
-        mySize += curMap->second.size();
-    }
-    mySize *= sizeof(uInt);
-    mySize += sizeof(Int) * fieldToSpwMap.size() + sizeof(uInt)*spwToFieldMap.size();
-    vector<std::set<Int> >::const_iterator map2End = spwToFieldMap.end();
-    uInt count = 0;
-    for (
-        vector<std::set<Int> >::const_iterator curMap = spwToFieldMap.begin();
-        curMap != map2End; ++curMap
-    ) {
-        count += curMap->size();
+    SHARED_PTR<const std::map<ScanKey, ScanProperties> > scanProps;
+    SHARED_PTR<const std::map<SubScanKey, SubScanProperties> > subScanProps;
+    _getScanAndSubScanProperties(scanProps, subScanProps, _showProgress);
+    std::map<SubScanKey, SubScanProperties>::const_iterator iter = subScanProps->begin();
+    std::map<SubScanKey, SubScanProperties>::const_iterator end = subScanProps->end();
+    for (; iter!=end; ++iter) {
+        Int fieldID = iter->first.fieldID;
+        const std::set<uInt>& spws = iter->second.spws;
+        if (fieldToSpwMap.find(fieldID) == fieldToSpwMap.end()) {
```

Comment:
this branch should not be necessary, the [] defaukt constructs a set on non-existance